878 link remapper method component defn

meganwolf0 opened the pull request:
## Description

Added a new method to component definition - `RewritePaths`
Intent of this method is to facilitate aggregation of data in component definitions (e.g., via Compose functionality) where the references contained therein may be relative to said component definition. This method should also have extensibility to be used for a future `packaging` type workflow where we may wish to pull disaggregated OSCAL models into a centralized, opinionated structure. Hence, this method may make sense to add to the interface when it's been implemented across models.

**Notes to reviewers**
* I updated the network function `GetLocalFileDir` - before it was doing a file existence check which I think is probably not needed in this function itself. Removal allowed for easier testing.
* This currently incorporates #747 and #748 since I needed the interface to be implemented in component definition, so those should probably be approved/merged first - this makes seeing the changes a little harder, but the main methods are in `src/pkg/common/common.go`

## Related Issue

Fixes #878 

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Other (security config, docs update, etc)

## Checklist before merging

- [ ] Test, docs, adr added or updated as needed
- [ ] [Schema Updates](https://github.com/defenseunicorns/lula/blob/main/docs/community-and-contribution/schema-updates.md) applied
- [ ] [Contributor Guide Steps](https://github.com/defenseunicorns/lula/blob/main/CONTRIBUTING.md) followed

mildwonkey commented on src/pkg/common/common.go:
```diff
@@ -211,3 +213,98 @@ func CleanMultilineString(str string) string {
+
+func splitPath(path string) []string {
+	components := strings.Split(path, ".")
+	return components
```
this can be a one-liner - probably not worth writing a method, but if you do, don't bother initializing the components variable, just return the output directly (`return strings.Split(path, "."`)

mildwonkey commented on src/pkg/common/common.go:
```diff
@@ -211,3 +213,98 @@ func CleanMultilineString(str string) string {
+	return reflectTraverseAndUpdate(reflect.ValueOf(obj), components, baseDir, newDir)
+}
+
+func reflectTraverseAndUpdate(val reflect.Value, components []string, baseDir string, newDir string) error {
```
This gives me a bit of anxiety - reflect is not a very performant package (and it moves more errors to runtime, away from compile time); it's something I generally avoid using in production code unless it's the only solutoin. That also doesn't mean it's not the correct choice, but it is something to be aware of. In the future we may want to add some benchmarks and keep an eye on how long it takes to process large documents.

mildwonkey commented on src/pkg/common/common_test.go:
```diff
@@ -529,6 +531,281 @@ func TestIsVersionValid(t *testing.T) {
+				},
+			},
+			curDir:        "/app",
+			newDir:        "/app/new",
```
In these examples `newDir` is a subdirectory of `curDir` - does this method work the same way if the `newDir` is 
 entirely different - `/otherapp`, or if curDir is `/app/new` and `newDir` is `/app`?

meganwolf0 replied:
It should work... I'll check this to see if there's a testing gap here

mildwonkey commented on src/pkg/common/oscal/component.go:
```diff
@@ -102,6 +102,51 @@ func (c *ComponentDefinition) HandleExisting(path string) error {
+	}
+
+	// Find all the paths in the component definition
+	pathsToRemap := []string{
```
This is a very interesting way of handling this. It feels string-and-reflect heavy to me. I don't know if I want to nitpick this as much as use it as the starting point when designing a larger refactor; we already have types for all of these, I thought, so why use reflection instead of walking the c.Model?

mildwonkey reviewed (commented):
I feel a bit uncomfortable with this implementation, but not so much that I'd block it - it's partly based on not entirely understanding what's going on in here and partly based on my entirely fuzzy mental image of a future refactor, which isn't exactly a valid consideration 😂

meganwolf0 commented:
> I feel a bit uncomfortable with this implementation, but not so much that I'd block it

Yeah I feel this too 😆 - I really wasn't sure _how_ to do this optimally/reusably...

So I think the bigger picture idea is that you can have an OSCAL model (this is just for component definition, but really any OSCAL Model) that can have links to other resources (remote or local). If we want to sort of package up all this data we may need to re-write the paths. Additionally, in the case of a component definition, it can import other component definitions (which contain these links) and those internal references will need to be re-written when the data gets imported to the importer model (ala, what we do with Compose right now, although those don't resolve paths, that function just puts all validations in the backmatter, which is not great for some of the use cases we've discussed with the SSP, etc).

The initial thought on this was to brute force look through all the spots where you'd possibly need to change a reference, and just do it. Although you can see from the list of items in `pathsToRemap` that it's a little on the long side. So totally agreed that reflect is not optimal, but that was my solution to not brute forcing it. Maybe there's another option that I'm overlooking if you have an idea on how to do it differently, I'm all ears!

mildwonkey commented:
My discomfort is because this feels like an implementation for iterating over raw json, but we have concrete types, so I'd expect something that iterated over each type.

Here's some probably-awful pseudo code:

func (c ComponentDefinition) RewritePaths (paths, anythingelse) error {
 for r := range c.Resources {
  if r.Href != "" {
    rewritePath(r.Href)
  }
  for s := range c.OtherField {
     // something like this for everything
  }
} 

Though for the actual implementation (and this is where my handwavey future ideas come into play, so again THIS MIGHT NOT MAKE SENSE for right now/ever) I'd probably have each type with links (every field inside the ComponentDefinition) fulfill an interface that includes a RewritePaths method, so we could handle each separately.

meganwolf0 commented:
Ok fair enough, yeah agreed maybe it would be better to just write a super long function and have all the types checked by type and not with reflect. This does get a little tedious as well because of all the pointers... To be honest, this is probably something I could coerce chat-gipity to give me so I didn't have to write it manually 🤣 

For the future idea, I truly don't think I know enough about go-oscal to know if that's do-able. But maybe worth investigating more...

I suppose for the near term we're probably trading the two options: using the reflect implementation or write a big function. (selfishly/lazily I do think this implementation is easier to test/understand)

meganwolf0 commented:
@brandtkeller if you have any weigh-ins on this TIA!

brandtkeller commented:
Reading back through the issue and then this implementation - I am weighing the "what problem does this solve now" with "are the surrounding assumptions correct". 

Yes the solutions complexity obfuscates some of the operations - I had to spend some time with debugging to get a clear-ish picture. 

But we have an original problem statement - composing component definitions that import other component definitions and remapping paths. Relevant problem for us - as each imported component-definition has potential links to Lula Validations or potentially other artifacts (which is why I appreciated the breadth of support in this PR). 

Apologies in advance. I believe we see the intent of a `package` operation for sake of transport - which allows us to dictate the artifact retrieval and replacement for re-mapping (evidence for a solution).

Do we see ourselves having any mechanism of getting away from a `compose` like operation? for component-definitions and allowing the import of other component-defintions - it feels like the answer is no and that this will be tied intimately to the execution of `validate` functionality - whether directly with Lula or with LaaL. Gut feeling here is that i'd default to a verbose solution that is easily traceable over a minimal solution that is less traceable.

meganwolf0 commented:
@brandtkeller @mildwonkey - Updated this PR to remove the reflect method (as that sounded broadly unpopular 😅 ) in favor of the verbose model walk/rewrite in place

meganwolf0 commented on src/pkg/common/oscal/component_test.go:
```diff
@@ -249,6 +249,7 @@ func TestMergeComponentDefinitions(t *testing.T) {
 		expectedImplementedRequirements       int
 		expectedTargetControlImplementations  int
 		expectedTargetImplementedRequirements int
+		uniqueComponent                       bool
```
oops this is what happens when I'm working on three branches at the same time, let me remove that and save for #893

brandtkeller reviewed (dismissed):
Thank you for the re-write - debugging traceability really improved my perspective. 

I'm a bit of a mess - this is all notional work not tied into execution currently - do we have issue(s) that add this to current execution?

brandtkeller commented:
> I'm a bit of a mess - this is all notional work not tied into execution currently - do we have issue(s) that add this to current execution?

#896 may be the execution implementation?

meganwolf0 commented:
> #896 may be the execution implementation

That's right - trying to decide if that PR should also include instrumenting the method in the generate SSP function or if that should be separate. But that's really the endgame here: enable the SSP having the right links to pass through to the Assessment Plan (so then we can do a thing with that).

meganwolf0 commented on src/pkg/common/oscal/component.go:
```diff
@@ -102,6 +102,279 @@ func (c *ComponentDefinition) HandleExisting(path string) error {
+	// BackMatter.Resources.Rlinks.Href
+	if c.Model.BackMatter != nil && c.Model.BackMatter.Resources != nil {
+		for _, resource := range *c.Model.BackMatter.Resources {
+			if resource.Rlinks != nil {
```
Looking at these a little more, I think the last one is not going to be possible unless we change go-oscal - and I think the types are auto-generated so honestly I think doing like an interface there would be not super straightforward.

I think there are a few types here which are redundant across models (e.g., BackMatter, Metadata, ResponsibleRoles, Links) that could be compartmentalized into reusable functions...

meganwolf0 commented:
Alright, I know you all already approved this but I think @mildwonkey was right about refactoring that huge function (ultimately I do see reuse across models, so the goal was to make that more reusable/clear/testable) - with that, I decided to not be lazy and made the following updates:
* Broke down the ComponentDefinition `RewritePaths` method into smaller functions (that I put in `common.go` because they aren't all specific to the component definition model (I didn't break out every type, the intent was to break into types that I saw reused and/or ones that were nested to facilitate testing)
* Updated the test files for ^^ method - honestly it seemed unnecessarily confusing with the abs paths in there and hopefully the point still stands. I think import/validate will use real paths but mock paths I believe are sufficient to test this. After going through the above exercise, I also wanted to address gaps in testing all the elements of ComponentDefinition
* Added a fuzz to `RemapPath`, hopefully it's a valid 😅

mildwonkey commented on src/pkg/common/common_test.go:
```diff
@@ -549,3 +656,11 @@ func FuzzReadValidationsFromYaml(f *testing.F) {
 	})
 }
+
+func FuzzRemapPath(f *testing.F) {
```
🎉 nice!!!

mildwonkey replied:
🤔 you may be able to round-trip this: (If this is valid) you could re-map back to the original path and add a check that last returned path is equal to the original test input. 

Something like this? 
gotPath := remap path (a, b, c) 
reversePath := (gotPath, c, b)

meganwolf0 replied:
Ok, I added this - it highlighted a thing where basically the path wasn't simplifying in the best way - e.g.,

If my starting path was ../app/file.txt, from the directory of /newapp, then you move it to /app, the file path still stayed the same (../app/file.txt) - which is _technically_ correct, however not really ideal. So I had to update the function. With this update, the test case "Invalid path remap" yields a different result... this is already an invalid path, but the way filepath.Join works it drops one of the "../", so the remap loses one as well... I'm kinda ok with this changing, since it was never a valid path, and since the purpose of this function isn't to actually stat the file, but just wanted to throw this out there that the change I made did affect tests and is probably a tradeoff of having a simplified path.

mildwonkey reviewed (commented):
I love the changes! I left one incredibly minor comment on `handlePath` but I think this is looking great.

meganwolf0 commented on src/pkg/common/network/network.go:
```diff
@@ -156,25 +156,45 @@ func FetchLocalFile(url *url.URL, config *fetchOpts) ([]byte, error) {
+	return handlePath(filepath.Join(url.Host, url.RequestURI()), baseDir)
+}
+
+func handlePath(path, baseDir string) string {
```
Long story short on this one - there's an option to network.Fetch (`WithBaseDir`) that requires having an empty directory to work with non-relative files (absolute paths or remote URLs). This is quite possibly a bad design with the functions (perhaps something in there should return like a bool if relative or something, then remove the option), however that's not what's happening. To change this would trickle a bit more than I wanted to with this PR, and it goes into compose territory, so I just renamed and commented this function.

I'm hoping that's good enough for this and thinking this is something we can look at properly cleaning up when the import component defns modifications (see #810) get properly folded in.

mildwonkey reviewed (approved):
❤️ :shipit:
